src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,380 @@
+*   limitations under the License.
+*/
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Compression;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Compression.Zstd
+{
+    internal class ZstandardStream : Stream
+    {
+        #region static
+        public static int MaxCompressionLevel => ZstandardNativeWrapper.MaxCompressionLevel; // maximum compression level available
+        #endregion
+
+        private readonly ArrayPool<byte> _arrayPool = ArrayPool<byte>.Shared;
+        private readonly Stream _compressedStream; // input for decompress and output for compress
+        private readonly int _defaultCompressionLevel = 6;
+
+        private readonly StreamReadHelper _streamReadHelper;
+        private readonly StreamWriteHelper _streamWriteHelper;
+        private readonly CompressionMode _compressionMode;
+
+        private bool _disposed;
+        private bool _flushed;
+        private readonly ZstandardNativeWrapper _nativeWrapper;
+
+        public ZstandardStream(
+            Stream compressedStream,
+            CompressionMode compressionMode,
+            Optional<int> compressionLevel = default)
+        {
+            _compressedStream = Ensure.IsNotNull(compressedStream, nameof(compressedStream));
+            _compressionMode = EnsureCompressionModeIsValid(compressionMode);
+
+            _nativeWrapper = new ZstandardNativeWrapper(_compressionMode, EnsureCompressionLevelIsValid(compressionLevel));
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    _streamWriteHelper = new StreamWriteHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedOutputSize));
+                    break;
+                case CompressionMode.Decompress:
+                    _streamReadHelper = new StreamReadHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedInputSize));
+                    break;
+            }
+        }
+
+        public override bool CanRead => _compressedStream.CanRead &&
+                                        _compressionMode == CompressionMode.Decompress;
+
+        public override bool CanWrite => _compressedStream.CanWrite &&
+                                         _compressionMode == CompressionMode.Compress;
+
+        public override bool CanSeek => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (!_disposed)
```

Comment:
Just an observation that the VS studio default style kind of hurts readability in cases like this :(